### ISSUE TITLE
BUG? Mockito evaluates a Function0 passed as argument

### ISSUE THREAD

**Issue by povilas** [etorreborre/specs2#629]
```scala
import org.specs2.mock.Mockito
import org.specs2.mutable.SpecWithJUnit

trait Foo {
  def withThunk[T](f: () => T): T
}

class FooTest extends SpecWithJUnit with Mockito {
  "foo" >> {
    val foo = mock[Foo]
    var calls = 0
    foo.withThunk(() => calls += 1)
    calls === 0
  }
}
```
Expected result: passing test
Actual result: `'2' is not equal to '0'`

Scala version: 2.12.4
Specs2 version: 3.8.6

**Comment by etorreborre** [comment]
This might be a consequence of specs2 trying to special-case by-name values with Mockito: https://github.com/etorreborre/specs2/blob/master/mock/shared/src/main/scala/org/specs2/mock/mockito/ArgumentsProcessor.scala#L54. The issue is that Mockito is a java library not really aware of Scala semantics :-(.

**Comment by etorreborre** [comment]
Is it possible for you to upgrade to a more recent version of specs2? In particular you might want to try `4.0.4-3593406-20180327185327` which supports the latest Mockito version.

**Comment by etorreborre** [comment]
Hi @povilas I am looking at this issue again and I don't know how to solve it. It might be possible to evaluate the thunk only once instead of 2 but due to how byname parameters are being checked with Mockito, I need to evaluate the thunk at least once to collect the possible matchers which would be used in place of a byname argument. So I am afraid this is going to be a "won't fix" issue :-(
